doc/develop/languages/cpp/index.rst

```diff
@@ -121,5 +121,133 @@ library can select, from its config,
+files are exposed to a much larger variety of toolchains and project
+configurations.
+Adding more constraints, the Zephyr project has demanding policies
+with respect to code style, compiler warnings, static analyzers and
+standard compliance (e.g.: MISRA).
+
+Put together, all these constraints can make writing header files very
+challenging. The purpose of this section is to document some best "header
+practices" and lessons learned in a Zephyr-specific context. While a lot
+of the information here is not Zephyr-specific, this section is not a
+substitute for knowledge of C/C++ standards, textbooks and other
+references.
+
+Testing
+-------
+
+Fortunately, the Zephyr project has an extensive test and CI
+infrastructure that provides coverage baselines, catches issues early,
+enforces policies and maintains such combinatorial explosions under some
+control. The ``tests/lib/cpp/cxx/`` are very useful in this context
+because their ``testcase.yaml`` configuration lets ``twister`` iterate
+quickly over a range of ``-std`` parameters: ``-std=c++98``,
+``-std=c++11``, etc.
+
+Keep in mind unused macros are not compiled.
+
+Designated initializers
+-----------------------
+
+Initialization macros are common in header files as they help reduce
+boilerplate code. C99 added initialization of ``struct`` and ``union``
+types by "designated" member names instead of a list of *bare*
+expressions. Some GCC versions support designated initializers even in
+their C90 mode.
+
+When used at a simple level, designated initializers are less
+error-prone, more readable and more flexible. On the other hand, C99
+allowed a surprisingly large and lax set of possibilities: designated
+initializers can be out of order, duplicated, "nested" (``.a.x =``),
+various other braces can be omitted, designated initializers and not can
+be mixed, etc.
+
+Twenty years later, C++20 added designated initializers to C++ but in
+much more restricted way; partly because a C++ ``struct`` is actually a
+``class``. As described in the C++ proposal number P0329 (which compares
+with C) or in any complete C++ reference, a mix is not allowed and
+initializers must be in order (gaps are allowed).
+
+Interestingly, the new restrictions in C++20 can cause ``gcc -std=c++20``
+to fail to compile code that successfully compiles with
+``gcc -std=c++17``. For example, ``gcc -std=c++17`` and older allow the
+C-style mix of initializers and bare expressions. This fails to compile
+with using ``gcc -std=c++20`` *with the same GCC version*.
+
+Recommendation: to maximize compatibility across different C and C++
+toolchains and standards, designated initializers in Zephyr header files
+should follow all C++20 rules and restrictions. Non-designated, pre-C99
+initialization offers more compatibility and is also allowed but
+designated initialization is the more readable and preferred code
+style. In any case, both styles must never be mixed in the same
+initializer.
```

Comment:
I think an example may help make this stick a bit better. That said I love the extensive explanation, if there's something I find irritating on our existing rules is how the misra links often give ZERO explanations of WHY one should follow those guidelines so good job on that front, it's just that I have a hard time visualizing the issue.

Same for the one below.

Reply:
I thought about examples but (as you seem to have noted) I was concerned about length: there are many rules and potential examples of what not to do. You could show only the one "correct" example but then it's not obvious what not to do. I actually gave one example: it's the git commit referenced at the bottom! It's short and the diff shows by essence both what to do and not. I could add more references to commits?

BTW there are some related fixes coming in #70403, so this will provide a couple more sample commits. This doc should be merged first though because it was requested in #70403 - circular dependency :-) I can add these commit examples to the doc later.

I didn't find many references to zephyr commits in the documentation so I'm not sure whether such references is a recommended documentation practice or not. I added the one at the bottom because it was "worth a thousand words" and saving as many. In general I like referencing commits because it's short, precise and 100% immutable which completely solves the big problem of documentation being often out of sync.

Reply:
Ok fair enough. I'll be totally honest, I think that at this point in the project any rule that is applied by humans rather than CI is not going to work. Too many guidelines, too many people, too many things to consider. If I look at myself for an example, I don't regularly work with C++ and haven't been bitten by any of these regularly by either projects or CI and just for that reason don't see myself remembering to check these two days from now. On the other side, stuff like the explicit casting for the `->api` void pointer is caught by CI and I find that that one managed to stick. Maybe we can have more combinations of C++ build? For different language versions or something. Anyway this is as good as it can be I suppose.

Reply:
> Ok fair enough. I'll be totally honest, I think that at this point in the project any rule that is applied by humans rather than CI is not going to work.

Exactly why #70403 is adding more test coverage.